src/TimeModeling/seismic_preconditioners.jl

```diff
@@ -161,6 +159,45 @@ end
 
 model_topmute(n::Tuple{Int64,Int64}, mute_end::Array{Float32, 2}, length, x) = vec(mute_end) .* vec(x)
 
+
+function model_topmute(n::Tuple{Int64,Int64,Int64}, mute_end::Array{Integer,2}, length::Int64, x_orig)
+    # Model domain topmute for a velocity model of dimensions n = [nx, nz].
+    x = deepcopy(reshape(x_orig,n))
+    for i=1:n[1]
+        for j=1:n[2]
+            mute_start = mute_end[i,j] - length
+            filter = zeros(Float32, n[end])
+            filter[1:mute_start-1] .= 0f0
+            filter[mute_end[i,j]+1:end] .= 1f0
+            taper_length = mute_end[i,j] - mute_start + 1
+            taper = (1f0 .+ sin.((Float32(pi)*(0:taper_length-1))/(taper_length - 1) .- Float32(pi)/2f0))/2f0
+            filter[mute_start:mute_end[i,j]] = taper
+            global x[i,j,:] = x[i,j,:].*filter
+        end
+    end
+    return vec(x)
+end
+    
+function model_topmute(n::Tuple{Int64,Int64,Int64}, mute_end::Int64, length::Int64, x_orig)
+    # Model domain topmute for a velocity model of dimensions n = [nx, nz].
+    x = deepcopy(reshape(x_orig,n))
+    mute_start = mute_end - length
+    filter = zeros(Float32, n[end])
+    filter[1:mute_start-1] .= 0f0
+    filter[mute_end+1:end] .= 1f0
+    taper_length = mute_end - mute_start + 1
+    taper = (1f0 .+ sin.((Float32(pi)*(0:taper_length-1))/(taper_length - 1) .- Float32(pi)/2f0))/2f0
```

Comment:
similar to water bottom 

```
    mute_start = mute_end - length
    filter = zeros(Float32, n[end])
    filter[1:mute_start-1] .= 0f0
    filter[mute_end+1:end] .= 1f0
    taper_length = mute_end - mute_start + 1
    taper = (1f0 .+ sin.((Float32(pi)*(0:taper_length-1))/(taper_length - 1) .- Float32(pi)/2f0))/2f0
``` 

these lines are repeated four times, that's not a good sign.

Reply:
Now I implemented 3D top mute based on 2D top mute in the latest commit. Thank you for your advice!